plugins/techdocs-addons-test-utils/src/test-utils.tsx

```diff
@@ -0,0 +1,240 @@
+import { screen } from 'testing-library__dom';
+import { renderToStaticMarkup } from 'react-dom/server';
+import { Route } from 'react-router-dom';
+import { act, render } from '@testing-library/react';
+
+import {
+  wrapInTestApp,
+  TestApiProvider,
+  TestApiProviderProps,
+} from '@backstage/test-utils';
+import { FlatRoutes } from '@backstage/core-app-api';
+
+import {
+  TechDocsAddons,
+  techdocsApiRef,
+  TechDocsEntityMetadata,
+  TechDocsMetadata,
+  techdocsStorageApiRef,
+} from '@backstage/plugin-techdocs-react';
+import { TechDocsReaderPage, techdocsPlugin } from '@backstage/plugin-techdocs';
+import { catalogPlugin } from '@backstage/plugin-catalog';
+import { searchApiRef } from '@backstage/plugin-search-react';
+
+/**
+ * @public
+ */
+export type RecursivePartial<T> = {
+  [P in keyof T]?: RecursivePartial<T[P]>;
+};
+
+const techdocsApi = {
+  getTechDocsMetadata: jest.fn(),
+  getEntityMetadata: jest.fn(),
+};
+
+const techdocsStorageApi = {
+  getApiOrigin: jest.fn(),
+  getEntityDocs: jest.fn(),
+  syncEntityDocs: jest.fn(),
+};
+
+const searchApi = {
+  query: jest.fn().mockResolvedValue({ results: [] }),
+};
+
+/**
+ * @public
+ */
+export type Apis = TestApiProviderProps<any>['apis'];
+
+type TechDocsAddonBuilderOptions = {
+  dom: ReactElement;
+  entity: RecursivePartial<TechDocsEntityMetadata>;
+  metadata: RecursivePartial<TechDocsMetadata>;
+  componentId: string;
+  apis: Apis;
+  path: string;
+};
+
+const defaultOptions: TechDocsAddonBuilderOptions = {
+  dom: <></>,
+  entity: {},
+  metadata: {},
+  componentId: 'docs',
+  apis: [],
+  path: '',
+};
+
+const defaultMetadata = {
+  site_name: 'Tech Docs',
+  site_description: 'Tech Docs',
+};
+
+const defaultEntity = {
+  kind: 'Component',
+  metadata: { namespace: 'default', name: 'docs' },
+};
+
+const defaultDom = (
+  <html lang="en">
+    <head />
+    <body>
+      <div data-md-component="container">
+        <div data-md-component="navigation" />
+        <div data-md-component="toc" />
+        <div data-md-component="main" />
+      </div>
+    </body>
+  </html>
+);
+
+/**
+ * @public
+ */
+
+export class TechDocsAddonBuilder {
+  private options: TechDocsAddonBuilderOptions = defaultOptions;
+  private addons: ReactElement[];
+
+  static buildAddonsInTechDocs(addons: ReactElement[]) {
+    return new TechDocsAddonBuilder(addons);
+  }
+
+  constructor(addons: ReactElement[]) {
+    this.addons = addons;
+  }
+
+  withApis(apis: Apis) {
+    const refs = apis.map(([ref]) => ref);
+    this.options.apis = this.options.apis
+      .filter(([ref]) => !refs.includes(ref))
+      .concat(apis);
+    return this;
+  }
+
+  withDom(dom: ReactElement) {
+    this.options.dom = dom;
+    return this;
+  }
+
+  withMetadata(metadata: RecursivePartial<TechDocsMetadata>) {
+    this.options.metadata = metadata;
+    return this;
+  }
+
+  withEntity(entity: RecursivePartial<TechDocsEntityMetadata>) {
```

Comment:
The entity doesn't seem to be merged together with any default or anything like that, so `RecursivePartial` here is a bit scary? Should probably be a `TechDocsEntityMetadata` directly instead?